File: python/oop2022/exs1/ex5.py
def read_int(name):
    while True:
        try:
            value = int(input("Insira o valor {}:\n".format(name)))
            return value
        except ValueError:
            print("Valor invalido inserido, tente novamente:")


def sqrt(x_0, error):
    guess = x_0 / 2

    while True:
        prev_guess = guess
        guess = (guess + x_0 / guess) / 2

        if abs(guess - prev_guess) < error:
            break

    return guess


def verify_primality(x):
    error = 0.00000001

    if x < 2:
        return False
    if x == 2:
        return True
    if x % 2 == 0:
        return False

    for i in range(3, int(sqrt(x, error)) + 1):
        if x % i == 0:
            return False
    return True


def main():
    x = read_int("do numero para achar o menor primo proximo")

    for i in range(x - 1, 2, -1):
        if verify_primality(i):
            print(f"{i} é o menor primo próximo de {x}")
            break


if __name__ == "__main__":
    main()

File: java/exercicio1/src/Ex5.java
import static java.lang.Math.abs;

public class Ex5 {

    // Function to obtain square root (^1/2) of given value, with maximum expected error
    static double sqrt(double x, double error) {

        double guess = x / 2, prev_guess;

        do {
            prev_guess = guess;
            guess = (guess + x / guess) / 2;
        } while (abs(guess - prev_guess) > error);

        return guess;
    }

    // Verify if all values up until square root of x divide x, else confirm that x is indeed prime
    static boolean is_prime(double x) {

        double error = 0.00000001;

        // Verify base cases
        if (x < 2)
            return false;
        if (x == 2)
            return true;
        if (x % 2 == 0)
            return false;

        for (int i = 3; i <= sqrt(x, error); i++) {
            if (x % i == 0)
                return false;
        }

        return true;
    }

    static int scan_int(String error) {

        int value;

        while (true) {
            try {
                value = EntradaTeclado.leInt();
                break;
            } catch (Exception e) {
                System.out.println(error);
            }
        }

        return value;
    }

    public static void main(String[] args) {

        System.out.println("Insira um numero para achar o menor primo mais proximo: ");

        int x = scan_int("Valor invalido inserido, tenhe novamente: ");

        boolean was_found = false;
        // Keep decrementing x until the next smallest prime number is found
        for (int i = x; i > 2; i--) {
            if (is_prime(i)) {
                System.out.printf("Resultado: %d", i);
                was_found = true;
                break;
            }
        }

        if (!was_found) {
            System.out.printf("Nenhum primo menor que %d foi encontrado\n", x);
        }
    }
}